File: greedy/maxProfit.java
/**
  卖股票问题
  Best Time to Buy and Sell Stock II
 
  Input: [7,1,5,3,6,4]
  Output: 7
  Explanation: 
    Buy on day 2 (price = 1) and sell on day 3 (price = 5), profit = 5-1 = 4.
    Then buy on day 4 (price = 3) and sell on day 5 (price = 6), profit = 6-3 = 3.
 */
 class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0)
            return 0;
            
        int maxprofit = 0;
        
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1])
                maxprofit += prices[i] - prices[i - 1];
        }
        return maxprofit;
    }
}

File: greedy/maxProfit.py
# -*- coding: utf-8 -*-

"""
使用贪心算法, 依次比较列表的相邻两个元素，如果后面的数大于前面的数，则买卖一次
对于 [7,1,5,3,6,4]
1，5可以倒卖获取利润4 
3，6可以倒卖获取利润3

对于[1,2,3,4,5]
每次倒卖获取利润1 最后总利润4
"""

class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        profit = 0
        for i, x in enumerate(prices):
            if i == len(prices) - 1: return profit
            # 遍历比较相邻两个数的值
            if x < prices[i+1]:
                profit += prices[i+1] - x
        return profit 

def main():
    prices = [7,1,5,3,6,4]
    ret = Solution().maxProfit(prices)
    print ret

if __name__ == '__main__':
    main()

      